lsp-types/src/Language/LSP/Types/Uri/OsPath.hs

```diff
@@ -0,0 +1,47 @@
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+#if MIN_VERSION_filepath(1,4,100)
+#define OS_PATH 1
+#endif
+
+module Language.LSP.Types.Uri.OsPath
+  (
+#ifdef OS_PATH
+    osPathToNormalizedFilePath
+  , normalizedFilePathToOsPath
+#endif
+  ) where
+
+#ifdef OS_PATH
+
+import           Control.DeepSeq        (NFData, force)
+import           Control.Exception      hiding (try)
+import           Control.Monad.Catch
+import           Language.LSP.Types.Uri
+import           System.IO.Unsafe       (unsafePerformIO)
+import           System.OsPath
+
+{-|
+Constructs 'NormalizedFilePath' from 'OsPath'. Throws 'IOException' if the conversion fails.
+-}
+osPathToNormalizedFilePath :: MonadThrow m => OsPath -> m NormalizedFilePath
+osPathToNormalizedFilePath = fmap toNormalizedFilePath . unsafePerformIO' . decodeFS
```

Comment:
I'm getting a little confused, would you mind writing a comment somewhere explaining the deal with the various encodings? The idea is something like: we always decode OsPaths using the system encoding into a string, and then we re-encode them as UTF8 to store them in NFP?

Reply:
I have missed all the discussion in the PR, so I need a comment explaining why we don't do the obvious thing (which would be storing the `OsPath` inside `NormalizedFilePath`).